File: 3027.java
import java.util.*;
public class Main {

 static public void main( String [] str ){

  Scanner sc=new Scanner(System.in);
  int tt=sc.nextInt();
  while(( tt--)!=0 ) {
    int a, b, r1, r2, t, r, s;
    a=sc.nextInt();
    b=sc.nextInt();
    r1=sc.nextInt();
    r2=sc.nextInt();
    for( r2--; r2>r1; r2-- ) {
	r = r2; t = 0; s = 1;
	while( r!=0 ) {
	  t += (r%a)*s;
	  s *= b;
	  r/=a;
	}
	if( t % r2 == 0 )
	  break;
    }
    if( r2 > r1 )
	System.out.printf( "%d\n", r2 );
    else
	System.out.printf( "Non-existent.\n" );
   }
  }
}

File: 3027.cc
#include <cstdlib>  
#include <cctype>  
#include <cstring>  
#include <cstdio>  
#include <cmath>  
#include <algorithm>  
#include <vector>  
#include <string>  
#include <iostream>  
#include <sstream>  
#include <map>  
#include <set>  
#include <queue>  
#include <stack>  
#include <fstream>  
#include <numeric>  
#include <iomanip>  
#include <bitset>  
#include <list>  
#include <stdexcept>  
#include <functional>  
#include <utility>  
#include <ctime>
using namespace std;
typedef long long LL;
typedef unsigned long long ULL;
#define MAX(a,b) ((a) > (b) ? (a) : (b))
#define MIN(a,b) ((a) < (b) ? (a) : (b))
#define MEM(a,b) memset((a),(b),sizeof(a))
const LL INF = 1e9;
const int N = 2e5;
LL check(int n, int b1, int b2)
{
	LL ret = 0;
	LL t = 1;
	int x = n;
	while (x)
	{
		ret += x%b1*t;
		x /= b1;
		t *= b2;
	}
	return ret%n == 0;
}
int main()
{
	int ncase;
	cin >> ncase;
	while (ncase--)
	{
		int b1, b2;
		int r1, r2;
		cin >> b1 >> b2 >> r1 >> r2;
		int ans = r1;
		for (int i = r1 + 1; i < r2; i++)
		{
			if (check(i, b1, b2))
			{
				ans = i;
			}
		}
		if (ans == r1) cout << "Non-existent." << endl;
		else cout << ans << endl;
	}
	return 0;

}



